SilverSolver/assignment4/islands.py

```diff
@@ -0,0 +1,36 @@
+from copy import deepcopy
+from typing import List
+
+def neighbourhood(i: int, j: int, islandMatrix: List[List[int]]) -> List[List[int]]:
+    "helper function, returns the neighbourhood of (i, j) cell in islandMatrix"
+
+    for di, dj in ((-1, 0), (1, 0), (0, -1), (0, 1)):
+        ni, nj = di + i, dj + j # ni, nj are a neighbor candidates for i, j
+        if 0 <= ni < len(islandMatrix) and 0 <= nj < len(islandMatrix[0]) and islandMatrix[ni][nj]:
+            yield ni, nj
+
+def island_fill(i: int, j: int, islandMatrix: List[List[int]]) -> List[List[int]]:
+    """
+    helper function, fills the island, to which (i, j) belongs, with "2"
+    it will help to distinct it with the islands in which we've never been
+    """
+    q = [(i, j)]
```

Comment:
Please rename q to something more meaningful, or write in the docstring the pseudo code of the algorithm :)

Do you mean 'queue'? Does this data structure holds the role of a 'visited' datastructure?

+    q = [(i, j)]
￼ +    while q:
￼ +        ei, ej = q.pop()
￼ +        if islandMatrix[ei][ej] == 1:
￼ +            for e in neighbourhood(ei, ej, islandMatrix):
￼ +                q.append(e)
￼ +            islandMatrix[ei][ej] = 2

One way to simplify this algorithm, IMHO is use the same datasctructure for q and islandMatrix. They mean almost the same thing: when q is empty, you have visited all the current island, and when islandMatrix has all the 1 turned into 2, you have also visited all the islands.